### ISSUE TITLE
ComponentRef Change detector doesn't trigger detecting for view

### ISSUE THREAD

**Issue by RolginRoman** [angular/angular#36667]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/core

### Is this a regression?

 Cannot say

### Description

 
**Component created by #ComponentFactoryResolver.**
ComponentFactoryResolver#createComponent returns ComponentRef that does have changeDetectorRef.   
**After any changes to component's input by code (not by template) and invocation of such ComponentRef#changeDetectorRef.detectChanges component's view won't be updated.**  

_example from repro: dynamic-wrapper.component.ts#createComponent._ 
> componentRef.changeDetectorRef.detectChanges()

This is root issue.

_To testing purposes ChangeDetectorRef was be added to component's constructor and invoked by 
consumer of this ComponentFactoryResolver. CDR from component's injector works as expected_ 

_example from repro: dynamic-wrapper.component.ts#createComponent_
> componentRef.instance.cdr.detectChanges()

## 🔬 Minimal Reproduction

 
[StackBlitz](https://stackblitz.com/github/RolginRoman/ng-dynamic-components-cdr-repro?file=src%2Fapp%2Fdynamic-wrapper%2Fdynamic-wrapper.component.ts)
  

GitHub repro repository:
https://github.com/RolginRoman/ng-dynamic-components-cdr-repro

## 🔥 Exception or Error
<pre><code>
No exception
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

 Angular version 9.1.0

</code></pre>

**Anything else relevant?**

Seems like browser, OS version or anything else doesn't matter

**Comment by pkozlowski-opensource** [comment]
This is a confirmed bug that still happens in the Angular v9 with ivy enabled (the same bug is present in VE). Here is a simplified reproduce scenario: https://stackblitz.com/edit/angular-i8tus5?file=package.json

What is going on here is that `changeDetectorRef` on a `ComponentRef` points to the change detector of the root (host) view of a dynamically created component. Then, inside the host view we've got the actual component view, but the component view is `OnPush` thus we never refresh it!

While it works as implemented (and as was implemented in VE), this behaviour is very confusing to the users and I would consider this to be broken from a user perspective. To actually understand why it doesn't work one needs to deeply understand inner workings of Angular (concept of the host view, OnPush etc.). I'm not sure what is the best long terms solution here but the fix won't be trivial.

The good news is that work-arrounds exists - the  simplest one would be to remove `ChangeDetectionStrategy.OnPush` from the dynamically created component - this is what I would recommend for now @RolginRoman .

**Comment by pkozlowski-opensource** [comment]
BTW, this issue will be very visible for people trying to test `OnPush` components - in fact this issue was already reported for `TestBed` as #12313 - it is currently highly up-voted and commented on issue biting people in practice.

**Comment by RolginRoman** [comment]
@pkozlowski-opensource thanks for your feedback! 🙂

Is it leads us to '_not to use onPush with dynamic components at all_'?
From my perspective such behaviour completely breaks many applications of dynamically-created components (from simple to advanced cases).
Component (or dev 😊) should be smart enough to understand where component will be created (from template or from such factories)

yeap, without `OnPush` it can work but it can affect app performance if there are many such components, can't it? Possible WA for it: detach CD from components and control CD cycles by self?

**Comment by montella1507** [comment]
I was researching a bug in our application for  5 days... Now i can confirm that ComponentRef.ChangeDetectorRef .detectChanges doesnt work, but   ComponentRef.instance.cdr (injected CDR via construcotr)  detectChanges works...

**Comment by montella1507** [comment]
Workarroud Is to use component's injector to get Real ChangeDetectorRef.

ComponentRef.injector.get(ChangeDetectorRef)

**Comment by jgomesmv** [comment]
I'm also facing this issue on angular 11.0.8. You can find a minimal reproduction here:
https://stackblitz.com/edit/angular-ivy-componentfactory-onpush-issue

Hope it helps!

**Comment by davoam** [comment]
> Workarroud Is to use component's injector to get Real ChangeDetectorRef.
> 
> ComponentRef.injector.get(ChangeDetectorRef)

In my case the following code helped
```
this.componentRef.injector.get(ChangeDetectorRef).detectChanges();
```

**Comment by hoeni** [comment]
Hell, this took my some time to find out this is an Angular issue (NG 13), not mine. But at least the workaround to get a fresh CDR from the injector worked well for me!

**Comment by montella1507** [comment]
@pkozlowski-opensource  as far as this bug is not yet fixed - can we atleast add note - warning, for example there:

https://angular.io/guide/dynamic-component-loader

that component's change detector is often not resolved 100% correctly  and its recommended to use components injector to resolve Change detector?

**Comment by pkozlowski-opensource** [comment]
This was just fixed via #46641 - the ComponentRef has a new method - `setInput` that correctly sets an input _and_ makes sure that change detection is properly invoked. More details in #46641
